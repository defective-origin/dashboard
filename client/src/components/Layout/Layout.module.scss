@import 'src/App/App.module';

@mixin init-markup() {
  grid-template-columns: auto 1fr auto;
  grid-template-rows: auto 1fr auto;
}

:global {
  .dark {
    // --text-color: white;
  }

  .light {
    // --text-color: black;
  }

  .layout {
    display: grid;

    // template
    &--cards {
      grid-auto-columns: 1fr;
      grid-auto-rows: 1fr;
    }

    &--row {
      grid-template-columns: repeat(1, 1fr);
      grid-auto-flow: column; /* Set the auto-placement algorithm to column */
      grid-auto-columns: minmax(100px, 1fr);
    }

    &--column {
      grid-template-columns: repeat(1, 1fr);
      grid-auto-flow: row; /* Set the auto-placement algorithm to column */
      grid-auto-rows: minmax(100px, 1fr);
    }

    &--rows {
      @include init-markup;

      grid-template-areas:
        'header header header'
        'left-aside content right-aside'
        'footer footer footer';
    }

    &--columns {
      @include init-markup;

      grid-template-areas:
        'header header header'
        'left-aside content right-aside'
        'footer footer footer';
    }

    &--header {
      @include init-markup;

      grid-template-areas:
        'header header header'
        'left-aside content right-aside'
        'left-aside footer right-aside';
    }

    &--footer {
      @include init-markup;

      grid-template-areas:
        'left-aside header right-aside'
        'left-aside content right-aside'
        'footer footer footer';
    }

    &--left-aside {
      @include init-markup;

      grid-template-areas:
        'left-aside header header'
        'left-aside content right-aside'
        'left-aside footer footer';
    }

    &--right-aside {
      @include init-markup;

      grid-template-areas:
        'header header right-aside'
        'left-aside content right-aside'
        'footer footer right-aside';
    }
  }
}
