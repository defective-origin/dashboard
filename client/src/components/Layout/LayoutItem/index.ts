export { default } from './LayoutItem.component'
export * from './LayoutItem.component'

export * from './LayoutItemUI'
export { default as useLayoutItemUIProps } from './LayoutItemUI'
