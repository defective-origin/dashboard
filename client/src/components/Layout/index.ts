/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Layout.component'

/* INJECT_EXPORT_PLACE */
export * from './Layout.component'
