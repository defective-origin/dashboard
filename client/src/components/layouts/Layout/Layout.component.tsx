import React from 'react'

// ---| core |---
import { react } from 'tools'

// ---| components |---
// ---| self |---
import useLayout, { LayoutOptions } from './Layout.hooks'

export type LayoutProps<E extends React.ElementType = React.ElementType> = react.CustomTagProps<LayoutOptions, E>

/**
 * Grid orientation component.
 * Allows to work with Grid items.
 *
 * Pay attention that preset components,
 * like: Content, Header and so on,
 * have block layout.
 *
 * How to use
 * @example
 * // grid markup
 * <Layout>
 *  <Item />
 *  <Item area='2/2' />
 *  <Item area='3/3' />
 *  <Block area='4/4' />
 *  <Layout area='5/5' />
 * </Layout>
 *
 * // markup
 * <Layout>
 *  <Item area='top' />
 *  <Item area='left' />
 *  <Item area='content' />
 *  <Block area='right' />
 *  <Layout area='bottom' />
 * </Layout>
 * // or via preset components
 * <Layout>
 *  <Header />
 *  <Left />
 *  <Content />
 *  <Right />
 *  <Footer />
 * </Layout>
 */
export function Layout<E extends React.ElementType = 'div'>(props: LayoutProps<E>) {
  const { as: Tag = 'div', ...layoutProps } = useLayout(props)

  return <Tag {...layoutProps} />
}

Layout.displayName = 'Layout'

export default Layout
