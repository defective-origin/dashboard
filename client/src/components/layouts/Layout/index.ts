/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Layout.component'

/* INJECT_EXPORT_PLACE */
export * from './Layout.component'
export * from './Layout.hooks'
