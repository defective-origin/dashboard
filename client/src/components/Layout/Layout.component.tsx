import React from 'react'

// ---| core |---
import { react } from 'tools'

// ---| components |---
import Block from 'components/Block'
import Section from 'components/Section'

// ---| self |---
import LayoutItem from './LayoutItem'
import useLayout, { LayoutOptions } from './Layout.hook'

export type LayoutProps = LayoutOptions

/**
 * Component description.
 *
 * How to use
 * @example
 * <Layout />
 */
export function Layout(props: LayoutProps): JSX.Element | null {
  const { children, className, style, ...otherProps } = useLayout(props)

  return (
    <div className={className} {...otherProps} style={style}>
      {children}
    </div>
  )
}

Layout.displayName = 'Layout'

export default react.attachComponents(Layout, {
  Item: LayoutItem,
  LeftAside: LayoutItem.LeftAside,
  RightAside: LayoutItem.RightAside,
  Footer: LayoutItem.Footer,
  Header: LayoutItem.Header,
  Content: LayoutItem.Content,
  Section: Section,
  Block: Block,
})
