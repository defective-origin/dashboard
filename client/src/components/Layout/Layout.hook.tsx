import React, { useMemo } from 'react'

// ---| core |---
import { cn } from 'tools'

// ---| pages |---
// ---| screens |---
// ---| components |---
import { BlockOptions, BlockReturnOptions, useBlock } from 'components/Block'

// ---| self |---
import './Layout.module.scss'

export type LayoutVariant = 'cards' | 'row' | 'rows' | 'column' | 'columns' | 'header' | 'footer' | 'left-aside' | 'right-aside' | 'grid'

export type LayoutOptions = BlockOptions & {
  v?: LayoutVariant
  areas?: React.CSSProperties['gridTemplateAreas']
  columns?: number,
  rows?: number,
}

export type LayoutReturnOptions<O extends object> = BlockReturnOptions<O>

/**
 * Hook descriptions
 *
 * @example
 * const options = useLayout(conf)
 */
export const useLayout = <O extends object>(options: LayoutOptions & O): LayoutReturnOptions<O> => {
  const { className, v = 'rows', areas, columns, rows, style, ...otherOptions } = useBlock(options)

  return useMemo(() => ({
    ...otherOptions,
    className: cn('layout', {
      [`layout--${v}`]: !areas,
    }, className),
    style: {
      gridTemplateAreas: areas,
      gridTemplateColumns: columns && `repeat(${columns}, 1fr)`,
      gridTemplateRows: rows && `repeat(${rows}, 1fr)`,
      ...style,
    },
  }) as LayoutReturnOptions<O>,
  [otherOptions, v, areas, className, columns, rows, style],
  )
}

export default useLayout
