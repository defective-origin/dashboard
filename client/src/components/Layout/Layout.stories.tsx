/* eslint-disable no-restricted-imports */
import { SB_CSS, field, params } from '../../../.storybook/tool'
import type { Meta, StoryObj } from '@storybook/react'
import { SIZES } from 'theme'
import Layout, { LayoutProps } from './Layout.component'
import Item from 'components/Item'
import Tag from 'components/Tag'

const CARD_VARIANTS: LayoutProps['v'][] = ['column', 'row', 'board']
const LAYOUT_VARIANTS: LayoutProps['v'][] = ['columns', 'rows', 'top', 'right', 'bottom', 'left']
const VARIANTS: LayoutProps['v'][] = [...CARD_VARIANTS, ...LAYOUT_VARIANTS]
const ALIGNS: LayoutProps['aligns'][] = ['start', 'center', 'end', 'baseline', 'stretch' ]
const JUSTIFIES: LayoutProps['justifies'][] = ['start', 'end', 'center', 'space-between', 'space-around', 'space-evenly']

const meta: Meta<typeof Layout> = {
  title: 'Components/LAYOUT/Layout',
  component: Layout,
  tags: ['autodocs'],
  argTypes: {
    className: field.string(),
    children: field.reactNode(),
    g: field.variants(SIZES, 'LayoutSpace'),
    p: field.variants(SIZES, 'LayoutSpace'),
    m: field.variants(SIZES, 'LayoutSpace'),
    v: field.variants(VARIANTS, 'LayoutDirection'),
    justifies: field.variants(JUSTIFIES, 'JustifyItems'),
    aligns: field.variants(ALIGNS, 'AlignItems'),
    grow: field.number('FlexGrow'),
    columns: field.number(),
    rows: field.number(),
    template: field.string(),
    cell: field.string(),
    flow: field.string('GridAutoFlow'),
    places: field.string('PlaceItems'),
    areas: field.string('GridTemplateAreas'),
  },
}

export default meta

type Story = StoryObj<typeof Layout>

const render = (props: LayoutProps) => (
  <Layout background={SB_CSS.margin}>
    {/* prevent margins from collapsing */}
    {props.m && <div style={{ height: '0.5px' }} />}

    <Layout minWidth={200} minHeight={200} background={SB_CSS.space} {...props}>
      {['left', 'right', 'top', 'bottom', 'center'].map((v) =>
        <Item
          key={v}
          background={SB_CSS.item}
          area={props.v && LAYOUT_VARIANTS.includes(props.v) ? v : undefined}
          minHeight={20}
          minWidth={20}
        />,
      )}
    </Layout>

    {/* prevent margins from collapsing */}
    {props.m && <div style={{ height: '0.5px' }} />}
  </Layout>
)

export const Demo: Story = {
  parameters: params('Layout [Requirements](?path=/docs/requirements-layout--docs)'),
  render,
  args: {
    v: 'columns',
    g: 'xxs',
    m: 'xxs',
    p: 'xxs',
  },
}

export const Variants: Story = {
  parameters: params('View', VARIANTS),
  render: () => (
    <Layout g='xs' v='board' columns={3} position='relative'>
      {VARIANTS.filter(Boolean).map((v) => (
        <Layout key={v} background={SB_CSS.space} v={v} g='xxs' p='xxs' columns={v === 'board' ? 2 : undefined} position='relative'>
          <Tag
            content={v}
            format='uppercase'
            style={{
              position: 'absolute',
              top: '50%',
              left: '50%',
              transform: 'translate(-50%, -50%)',
            }}
          />

          {['left', 'right', 'top', 'bottom', 'center'].map((i) =>
            <Item
              key={i}
              background={SB_CSS.item}
              area={LAYOUT_VARIANTS.includes(v) ? i : undefined}
              minHeight={50}
              minWidth={50}
            />,
          )}
        </Layout>
      ))}
    </Layout>
  ),
}

export const Spaces: Story = {
  parameters: params('Margin[m] | Padding[p] | Gap[g]', SIZES),
  render,
  args: {
    v: 'columns',
    g: 'xxs',
    m: 'xxs',
    p: 'xxs',
  },
}

export const Justifies: Story = {
  parameters: params('Justify', JUSTIFIES),
  render,
  args: {
    v: 'columns',
    g: 'xxs',
    justifies: 'center',
  },
}

export const Aligns: Story = {
  parameters: params('Align', ALIGNS),
  render,
  args: {
    v: 'columns',
    g: 'xxs',
    aligns: 'center',
  },
}
