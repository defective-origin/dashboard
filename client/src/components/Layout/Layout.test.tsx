// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Layout from './Layout.component'

describe('[Layout] component', () => {
  it('should render component', () => {
    const container = render(<Layout />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})
