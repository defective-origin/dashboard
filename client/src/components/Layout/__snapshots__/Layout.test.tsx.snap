// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Layout] component > should render component 1`] = `
<DocumentFragment>
  <div
    class="layout layout--row"
  >
    <div
      class="layout-item layout-item--content"
    />
  </div>
</DocumentFragment>
`;
